### PR TITLE
[coldstart][dropoutnet][MetricLearningI2I]Add Cold start model dropoutnet，Add Collaborative Metric Learning i2i model

### PR DESCRIPTION
**Pull request by yangxudong**
- Add cold start model  DropoutNet
- Add Collaborative Metric Learning i2i model
- Add TextCNN model

### CODE REVIEW

File: easy_rec/python/model/easy_rec_model.py [hunk 1]
```diff
@@ -50,7 +50,11 @@ def __init__(self,
 
     # add sample weight from inputs
     self._sample_weight = 1.0
-    if constant.SAMPLE_WEIGHT in features:
+    if self._model_config.HasField('sample_weight_field'):
```
Comment (chengmengli06) [inline easy_rec/python/model/easy_rec_model.py:53]: why another sample weight field?

File: docs/source/models/dropoutnet.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  embedding_regularization: 5e-5
+}
+```
+
+- model_class: 'DropoutNet', 不需要修改
+- feature_groups: 需要四个feature_group: user_content、user_preference和item_content、item_preference, **group name不能变**。
+  其中，user_content和user_preference两者至少要有1个；item_content和item_preference两者至少要有1个。
+- losses: 配置模型使用的loss function的类型和权重；DropoutNet支持三种类型的loss函数，可以单独使用，也可以搭配在一起使用。
+  - loss_type=CLASSIFICATION: 常规的二分类sigmoid cross entropy loss
+  - loss_type=PAIR_WISE_LOSS: pairwise rank loss，也就AUC loss，以优化全局AUC为目标；
```
Comment (poson) [inline docs/source/models/dropoutnet.md:96]: 也就是AUC loss

File: docs/source/models/textcnn.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# TextCNN
+
+### 简介
+
+TextCNN网络是2014年提出的用来做文本分类的卷积神经网络，由于其结构简单、效果好，在文本分类、推荐等NLP领域应用广泛。
+对于文本分类问题，常见的方法无非就是抽取文本的特征。然后再基于抽取的特征训练一个分类器。 然而研究证明，TextCnn在文本分类问题上有着更加卓越的表现。
```
Comment (chengmengli06) [inline docs/source/models/textcnn.md:6]: how it is related to recommendation? add some explanations

File: easy_rec/python/utils/test_utils.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 from easy_rec.python.utils import config_util
 from easy_rec.python.protos.pipeline_pb2 import EasyRecConfig
 
-TEST_DIR = '/tmp/easy_rec_test'
+TEST_DIR = './tmp/easy_rec_test'
```
Comment (chengmengli06) [inline easy_rec/python/utils/test_utils.py:27]: it is better not to update this

File: samples/model_config/text_cnn.config [hunk 5]
```diff
@@ -0,0 +1,85 @@
+train_input_path: "data/test/tb_data/taobao_train_data"
+eval_input_path: "data/test/tb_data/taobao_test_data"
```
Comment (dawn310826) [inline samples/model_config/text_cnn.config:2]: 这个数据没有titile，是如何跑通的？如果有新的实验数据，可以提交到oss上。

File: easy_rec/python/utils/test_utils.py (hunk 4 above)
Comment (yangxudong) [reply easy_rec/python/utils/test_utils.py:27]: 在很多机器上往/tmp目前下写数据是需要超级用户权限的，我们有时候很难申请到超级用户权限；测试不方便。

File: easy_rec/python/model/dropoutnet.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        loss_value = tf.losses.sigmoid_cross_entropy(
+          labels,
+          logits,
+          self._sample_weight)
+        self._loss_dict['sigmoid_loss'] = loss_value * loss.weight
+    return self._loss_dict
+
+  def build_metric_graph(self, eval_config):
+    metric_dict = {}
+    label = list(self._labels.values())[0]
```
Comment (poson) [inline easy_rec/python/model/dropoutnet.py:193]: 上面都是labels ，有点不一致

File: easy_rec/python/model/dropoutnet.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            labels,
+            self._model_config.softmax_loss.num_negative_samples,
+            embed_normed=True,
+            weights=self._sample_weight,
+            margin=self._model_config.softmax_loss.margin,
+            gamma=self._model_config.softmax_loss.gamma,
+            t=self._model_config.softmax_loss.coefficient_of_support_vector)
+        self._loss_dict['softmax_loss'] = loss_value * loss.weight
+      elif loss.loss_type == LossType.PAIR_WISE_LOSS:
+        loss_value = pairwise_loss(logits, labels)
```
Comment (poson) [inline easy_rec/python/model/dropoutnet.py:181]: 要不要保持labels 在前面

File: easy_rec/python/model/easy_rec_model.py (hunk 1 above)
Comment (yangxudong) [reply easy_rec/python/model/easy_rec_model.py]: 已经删了

File: samples/model_config/text_cnn.config (hunk 5 above)
Comment (yangxudong) [reply samples/model_config/text_cnn.config]: TextCNN model 删掉了

File: easy_rec/python/model/dropoutnet.py (hunk 7 above)
Comment (yangxudong) [reply easy_rec/python/model/dropoutnet.py]: 已修改

File: easy_rec/python/model/dropoutnet.py (hunk 6 above)
Comment (yangxudong) [reply easy_rec/python/model/dropoutnet.py]: 已修改

File: easy_rec/python/core/metrics.py [hunk 8]
```diff
@@ -127,3 +129,105 @@ def session_auc(labels, predictions, session_ids, reduction='mean'):
...
+                                labels,
+                                session_ids=None,
+                                embed_normed=False):
+  """Computes the recall_at_k metric for metric learning.
+
+  Args:
+    embeddings: the output of last hidden layer, a tf.float32 `Tensor` with shape [batch_size, embedding_size]
+    session_ids: session ids, a `Tensor` with shape [batch_size]
+    labels: a `Tensor` with shape [batch_size]
+    k: a scalar of int, or a tuple of ints
```
[the remark this answers is not in the document]
Comment (yangxudong) [reply easy_rec/python/core/metrics.py]: 已修改

File: docs/source/models/textcnn.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+从直观上理解，TextCNN通过一维卷积来获取句子中n-gram的特征表示。
+TextCNN对文本浅层特征的抽取能力很强，在短文本领域专注于意图分类时效果很好，应用广泛，且速度较快。
+
+Yoon Kim在论文\[EMNLP 2014\][Convolutional neural networks for sentence classication](https://www.aclweb.org/anthology/D14-1181.pdf)
+提出了TextCNN并给出基本的结构。
+将卷积神经网络CNN应用到文本分类任务，利用多个不同size的kernel来提取句子中的关键信息（类似于多窗口大小的ngram），从而能够更好地捕捉局部相关性。
```
[the remark this answers is not in the document]
Comment (yangxudong) [reply docs/source/models/textcnn.md]: 已删除

File: docs/source/models/textcnn.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (yangxudong) [reply docs/source/models/textcnn.md]: 已删除

**Review by poson — APPROVED** [review]
